presto-raptor/src/test/java/com/facebook/presto/raptor/storage/TestOrcFileRewriter.java

```diff
@@ -413,9 +428,225 @@ public void testUncompressedSize(boolean useOptimizedOrcWriter)
+    public void testRewriterDropThenAddDifferentColumns(boolean useOptimizedOrcWriter)
+            throws Exception
+    {
+        TypeRegistry typeRegistry = new TypeRegistry();
+        DBI dbi = new DBI("jdbc:h2:mem:test" + System.nanoTime());
+        dbi.registerMapper(new TableColumn.Mapper(typeRegistry));
+        Handle dummyHandle = dbi.open();
+        File dataDir = Files.createTempDir();
+
+        StorageManager storageManager = createOrcStorageManager(dbi, dataDir, useOptimizedOrcWriter);
+
+        List<Long> columnIds = ImmutableList.of(3L, 7L);
+        List<Type> columnTypes = ImmutableList.of(BIGINT, createVarcharType(20));
+
+        File file = new File(temporary, randomUUID().toString());
+        try (FileWriter writer = createFileWriter(columnIds, columnTypes, file, false, useOptimizedOrcWriter)) {
+            List<Page> pages = rowPagesBuilder(columnTypes)
+                    .row(2L, "2")
+                    .build();
+            writer.appendPages(pages);
+        }
+
+        // Add a column
+        File newFile1 = new File(temporary, randomUUID().toString());
+        OrcFileInfo info = createFileRewriter(useOptimizedOrcWriter).rewrite(
+                getColumnTypes(ImmutableList.of(3L, 7L, 10L), ImmutableList.of(BIGINT, createVarcharType(20), DOUBLE)),
+                file,
+                newFile1,
+                new BitSet(5));
+        assertEquals(info.getRowCount(), 1);
+
+        // Drop a column
+        File newFile2 = new File(temporary, randomUUID().toString());
+        info = createFileRewriter(useOptimizedOrcWriter).rewrite(
+                getColumnTypes(ImmutableList.of(7L, 10L), ImmutableList.of(createVarcharType(20), DOUBLE)),
+                newFile1,
+                newFile2,
+                new BitSet(5));
+        assertEquals(info.getRowCount(), 1);
+
+        // Add a column with the different ID with different type
+        File newFile3 = new File(temporary, randomUUID().toString());
+        info = createFileRewriter(useOptimizedOrcWriter).rewrite(
+                getColumnTypes(ImmutableList.of(7L, 10L, 13L), ImmutableList.of(createVarcharType(20), DOUBLE, createVarcharType(5))),
+                newFile2,
+                newFile3,
+                new BitSet(5));
+        assertEquals(info.getRowCount(), 1);
+
+        // Get prepared for the final file; make sure it is accessible from storage manager
+        UUID uuid = randomUUID();
+        File newFile4 = getFileSystemPath(new File(dataDir, "data/storage"), uuid);
+        if (useOptimizedOrcWriter) {
+            // Apache ORC writer creates the file itself
+            newFile4.getParentFile().mkdirs();
+            newFile4.createNewFile();
+        }
+
+        // Drop a column and add a column
+        info = createFileRewriter(useOptimizedOrcWriter).rewrite(
+                getColumnTypes(ImmutableList.of(7L, 13L, 8L), ImmutableList.of(createVarcharType(20), createVarcharType(5), INTEGER)),
```

Comment:
In practice it's impossible to have a new column whose column id is smaller. 

In Raptor, we refuse dropping the column that has the max column id. And when adding column, we use max column id plus 1. In that way, we never reuse column ids.

I admit that it's not reasonable to refuse dropping the last column. We can fix it by changing the metadata schema. Instead of keeping track of current max column id when creating new column, we keep track of the historical max column id in the table metadata and use that when creating new column.